server/team_matchmaking_service.py

```diff
@@ -0,0 +1,115 @@
+import time
+from typing import NamedTuple
+
+from .game_service import GameService
+from .decorators import with_logger
+from server.lobbyconnection import ClientError
+from .players import Player
+from .team_matchmaker.player_party import PlayerParty
+
+MapDescription = NamedTuple('Map', [("id", int), ("name", str), ("path", str)])
+GroupInvite = NamedTuple('GroupInvite', [("sender", Player), ("recipient", Player), ("party", PlayerParty), ("created_at", float)])
+
+PARTY_INVITE_TIMEOUT = 60 * 60 * 24  # secs
+
+@with_logger
+class TeamMatchmakingService:
+    """
+    Service responsible for managing the global team matchmaking. Does grouping, matchmaking, updates statistics, and
+    launches the games.
+    """
+
+
+    def __init__(self, games_service: GameService):
+        self.game_service = games_service
+        self.player_parties: dict[Player, PlayerParty] = dict()
+        self._pending_invites: dict[(Player, Player), GroupInvite] = dict()
+
+    def invite_player_to_party(self, sender: Player, recipient: Player):
+        if sender not in self.player_parties:
+            self.player_parties[sender] = PlayerParty(self, sender)
+
+        party = self.player_parties.get(sender)
+
+        if not party.owner == sender:
+            raise ClientError("You do not own this party.", recoverable=True)
+
+        if sender in recipient.foes:
+            raise ClientError("This person doesn't accept invites from you.", recoverable=True)
+
+        self._pending_invites[(sender, recipient)] = GroupInvite(sender, recipient, party, time.time())
+        recipient.lobby_connection.send({
+            "command": "party_invite",
+            "sender": sender.id
+        })
+
+    def accept_invite(self, recipient: Player, sender: Player):
+        if (sender, recipient) not in self._pending_invites:
+            raise ClientError("You're not invited to a party", recoverable=True)
+
+        pending_invite = self._pending_invites.pop((sender, recipient))
+
+        if pending_invite.party not in self.player_parties:
+            raise ClientError("The party you're trying to join doesn't exist anymore.", recoverable=True)
+
+        pending_invite.party.add_player(recipient)
+
+        self.remove_disbanded_parties()
+
+    def kick_player_from_party(self, owner: Player, kicked_player: Player):
+        if owner not in self.player_parties:
+            raise ClientError("You're not in a party.", recoverable=True)
+
+        party = self.player_parties.get(owner)
```

Comment:
Can use `self.player_parties[owner]` here. When I see `.get()` that usually implies to me that we are expecting the possibility that the key does not exist, however, that isn't the case here since the code will already have errored out if the key doesn't exist.